source/common/grpc/typed_async_client.h

```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include <chrono>
+
+#include "envoy/grpc/async_client.h"
+
+namespace Envoy {
+namespace Grpc {
+
+/**
+ * Convenience wrapper for an AsyncStream* providing typed protobuf support.
+ */
+class UntypedAsyncStream {
+public:
+  UntypedAsyncStream() {}
+  UntypedAsyncStream(AsyncStream* stream, bool is_grpc_header_required)
+      : stream_(stream), is_grpc_header_required_(is_grpc_header_required) {}
+  void sendMessage(Buffer::InstancePtr&& request, bool end_stream) {
+    stream_->sendMessage(std::move(request), end_stream);
+  }
+  void closeStream() { stream_->closeStream(); }
+  void resetStream() { stream_->resetStream(); }
+  void sendMessageUntyped(const Protobuf::Message& request, bool end_stream);
+  UntypedAsyncStream* operator->() { return this; }
+  bool operator==(AsyncStream* stream) const { return stream_ == stream; }
+  bool operator!=(AsyncStream* stream) const { return stream_ != stream; }
+
+private:
+  template <typename Request> friend class TypedAsyncStream;
+  AsyncStream* stream_{};
+  bool is_grpc_header_required_{};
+};
+
+template <typename Request> class TypedAsyncStream : public UntypedAsyncStream {
```

Comment:
What is the benefit of having two classes here? The untyped one is actually just a proto wrapper around the real untyped async client (and is somewhat typed, it has a proto interface), then we have the typed wrapper which doesn't seem to do very much. It seems we could collapse the two together?

Reply:
This was the format that we had before.  I am happy to get rid of it, but I was trying to preserve the original behavior.

Reply:
Done. PTAL